src/porepy/numerics/ad/grid_operators.py

```diff
@@ -557,6 +563,86 @@ def __str__(self) -> str:
         return s
 
 
+class Geometry(Operator):
+    """Wrapper class for Ad representations of grids.
+
+
+    Attributes:
+        cell_volumes (pp.ad.Matrix): Diagonal ad matrix of cell volumes.
+        face_areas (pp.ad.Matrix):  Diagonal ad matrix of face areas.
+        nd (int): Ambient/highest dimension of the mixed-dimensional grid.
+
+    FIXME: Implement parse??
+    """
+
+    def __init__(self, subdomains: List[pp.Grid], nd: int, name: Optional[str] = None):
+        """Construct concatenated grid operators for a given set of subdomains.
+
+        The operators will be ordered according to the ordering in grids. It is critical
+        that the same ordering is used by other operators.
+
+        Parameters:
+            subdomains (List of pp.Grid): List of grids. The order of the grids in the list
+                sets the ordering of the geometry operators. Can be either subdomain (pp.Grid)
+                or interface (pp.MortarGrid) grids.
+            nd: ambient dimension.
+            name (str, optional): Name of the operator. Default is None.
+
+        """
+        super().__init__(name=name)
+
+        self.subdomains = subdomains
+        self._num_grids: int = len(subdomains)
+        self.nd = nd
+
+        self.num_cells: int = sum([g.num_cells for g in subdomains])
+        self.num_faces: int = sum([g.num_faces for g in subdomains])
+
+        # Wrap the stacked matrices into Ad objects (could be extended to e.g. face normals)
+        for field in ["cell_volumes", "face_areas"]:
+            ad_matrix = Matrix(
+                sps.diags(np.hstack([getattr(g, field) for g in subdomains]))
+            )
+            setattr(self, field, ad_matrix)
+
+        def scalar_to_nd(size):
+            """Expand matrix of size [N x M] to [nd*N x M].
+
+            When left multiplied to a matrix A, each row of A is
+            repeated nd times.
+
+            Usage example: Scaling from traction to force is
+
+                    force = (scalar_to_nd_face * face_areas) * traction
+
+            FIXME: Refactor?
```

Comment:
Come to think of it, I believe this is very similar to what we do here https://github.com/pmgbergen/porepy/blob/83c2e4af5094cf3e445d2e3db2274493a1a3c14e/src/porepy/numerics/fv/fvutils.py#L433 (or perhaps in the function below that one), although the present function of course has an Ad wrapping.

Suggested solution: Keep this for now, revisit at a later point (while refactoring parts of fvutils into array_operations).